applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -0,0 +1,888 @@
+            r_element.Set(TO_ERASE, true);
+        }
+        for (auto& r_node : wake_sub_model_part.Nodes()){
+            r_node.Set(TO_ERASE, true);
+        }
+        wake_sub_model_part.RemoveElements(TO_ERASE);
+        wake_sub_model_part.RemoveNodes(TO_ERASE);
+    }
+    else{
+        // Creating the wake_sub_model_part
+        root_model_part.CreateSubModelPart("wake_elements_model_part");
+    }
+}
+
+// This function marks the trailing edge nodes and finds the wingtip nodes
+void Define3DWakeProcess::MarkTrailingEdgeNodesAndFindWingtipNodes()
+{
+    double max_span_position = std::numeric_limits<double>::lowest();
+    double min_span_position = std::numeric_limits<double>::max();
+
+    auto p_right_wing_tip_node = &*mrTrailingEdgeModelPart.NodesBegin();
+    auto p_left_wing_tip_node = &*mrTrailingEdgeModelPart.NodesBegin();
+
+    for (auto& r_node : mrTrailingEdgeModelPart.Nodes()) {
+        r_node.SetValue(TRAILING_EDGE, true);
+        const auto& r_coordinates = r_node.Coordinates();
+        const double distance_projection = inner_prod(r_coordinates, mSpanDirection);
+
+        // Wingtip nodes have maximum and minimum span positions
+        if(distance_projection > max_span_position){
+            p_right_wing_tip_node = &r_node;
+            max_span_position = distance_projection;
+        }
+        if(distance_projection < min_span_position){
+            p_left_wing_tip_node = &r_node;
+            min_span_position = distance_projection;
+        }
+    }
+
+    // Marking wingtip nodes
+    p_right_wing_tip_node->SetValue(WING_TIP, true);
+    p_left_wing_tip_node->SetValue(WING_TIP, true);
+}
+
+// This function computes the wing lower surface normals and marks the upper
+// and lower surfaces. The wing lower surface normals are used later in
+// RecomputeComputeNodalDistancesToWakeOrWingLowerSurface inside the
+// MarkKuttaElements function to check whether nodes are above or below the wake
+// TODO: Think a better way of doing this.
+void Define3DWakeProcess::ComputeWingLowerSurfaceNormals() const
+{
+    // TO DISCUSS: Is it worth it to run these loops in parallel?
+    // So far it has not been noted much difference in terms of speed.
+    // Also, when ran in parallel, the result can be random.
+    // Mark upper surface
+    for (auto& r_cond : mrBodyModelPart.Conditions()) {
+        auto& r_geometry = r_cond.GetGeometry();
+        const auto& surface_normal = r_geometry.UnitNormal(0);
+        const double projection = inner_prod(surface_normal, mWakeNormal);
+
+        if(!(projection > 0.0)){
+            for (unsigned int j = 0; j < r_geometry.size(); j++) {
+                r_geometry[j].SetLock();
+                // QUESTION: Do we need to initialize the variable UPPER_SURFACE to false?
+                r_geometry[j].SetValue(UPPER_SURFACE, true);
+                r_geometry[j].UnSetLock();
+            }
+        }
+    }
```

Comment:
parallel utils

Reply:
When ran in parallel the result is random, I need to think of a better way of doing this. Any suggestions are welcome.